### DISCUSSION TITLE
Is the concept of SQL4CDSConnection MaxDegreeOfParallelism same with ServiceClient RecommendedDegreesOfParallelism?

### DISCUSSION THREAD

**Discussion by jarvan-jiang** [MarkMpn/Sql4Cds#529] [Q&A] [answered]
I am so trilled to see it is possible to use `SQL4CDSConnection` to perform data operations!
In the old times of using `ServiceClient`, to get better performance usually we will use `RecommendedDegreesOfParallelism` which returned from server, to execute parallel fetch or CRUD requests with the recommend parallelism number .
However, I noticed in `SQL4CDSConnection`, the prop `MaxDegreeOfParallelism` is set manually, and not inherit from `ServiceClient`. And it seems `SQL4CDSConnection` will automatically optimize query to parallel. 
So I am wondering should I set the value from `ServiceClient` manually? Or they are complete different concepts, if so, how can we chose a proper value for different environments?

**Comment by MarkMpn** [comment, marked as answer]
If you set the `MaxDegreeOfParallelism` property to 0, it will use the `RecommendedDegreesOfParallelism` for any DML operations.  If you set it to any non-zero value it will use that instead and ignore the value from the ServiceClient.

Fetch operations are not executed in parallel as it needs to get the results of one page in order to get the paging cookie to efficiently retrieve the next page.
